File: backend/emailgen/image_generator.py
"""图片生成模块

双轨策略：
1. **wanx_enabled**：当 config.qianwen_vision 配置了 api_key + base_url 时，调用阿里万相
   （Token Plan 兼容模式 /chat/completions 多模态）生成高质量定制营销图。
2. **Fallback**：若未配置万相，用原版 Pollinations（免 Key）生成通用产品图，再用 Pillow 叠加
   折扣文字/CTA 按钮。保证零配置也能出图，不阻塞草稿生成。
"""
from __future__ import annotations

import os
import re
import time
from pathlib import Path
from typing import Any, Dict

import requests

from .config import Config
from .copy_generator import generate_image_prompt
from .data_loader import UserRecord
from .image_overlay import overlay_marketing_text


# ---------------------------------------------------------------------------
# 万相（Token Plan 兼容模式）
# ---------------------------------------------------------------------------


def _extract_image_url(data: Dict[str, Any]) -> str:
    out = data.get("output") or {}
    choices = out.get("choices") or data.get("choices") or []
    if not choices:
        return ""
    msg = choices[0].get("message") or {}
    content = msg.get("content")

    if isinstance(content, list):
        for part in content:
            if not isinstance(part, dict):
                continue
            if part.get("image"):
                return part["image"]
            if part.get("type") == "image_url" and isinstance(part.get("image_url"), dict):
                return part["image_url"].get("url", "")
            if part.get("url"):
                return part["url"]
        return ""

    if isinstance(content, str):
        m = re.search(r"\((https?://[^)]+)\)", content) or re.search(r"(https?://\S+)", content)
        return m.group(1) if m else ""

    return ""


def _request_wanx(prompt: str, config: Config) -> str:
    q = config.qianwen_vision
    base = (q.base_url or "").rstrip("/")
    if not base or not q.api_key:
        raise RuntimeError("万相未配置（缺少 base_url 或 api_key）")
    url = f"{base}/chat/completions"
    headers = {
        "Authorization": f"Bearer {q.api_key}",
        "Content-Type": "application/json",
    }
    payload = {
        "model": q.model,
        "messages": [
            {"role": "user", "content": [{"type": "text", "text": prompt}]},
        ],
    }
    resp = requests.post(url, headers=headers, json=payload, timeout=180)
    if resp.status_code != 200:
        raise RuntimeError(f"万相 HTTP {resp.status_code}: {resp.text[:400]}")
    data = resp.json()
    image_url = _extract_image_url(data)
    if not image_url:
        raise RuntimeError(f"万相响应无图片 URL: {str(data)[:400]}")
    return image_url


# ---------------------------------------------------------------------------
# Pollinations 免 Key 兜底
# ---------------------------------------------------------------------------


def _request_pollinations(discount: float, brand: str, seed_salt: str = "") -> str:
    prompt = (
        "High-quality e-commerce marketing email hero image of a premium phone case, "
        "sleek modern smartphone case product photography, "
        "bright clean studio lighting, soft gradient background, "
        "professional brand aesthetic, newsletter banner style, "
        "8K detailed, no text overlay in the image, best quality"
    )
    encoded = requests.utils.quote(prompt)
    seed = (int(time.time()) + abs(hash((brand, discount, seed_salt))) % 100_000) % 1_000_000
    return (
        f"https://image.pollinations.ai/prompt/{encoded}"
        f"?width=1200&height=1500&seed={seed}&nologo=true"
    )


# ---------------------------------------------------------------------------
# 下载 / 入口
# ---------------------------------------------------------------------------


def download_image(image_url: str, tag: str, output_dir: str = "output/images", timeout: int = 90) -> str:
    """下载远程图片到本地 output/images/，返回绝对路径；失败返回空串"""
    out_dir = Path(output_dir)
    out_dir.mkdir(parents=True, exist_ok=True)
    safe_tag = re.sub(r"[^\w.-]+", "_", str(tag))[:60] or "img"
    filename = f"{safe_tag}_{int(time.time())}.png"
    local_path = out_dir / filename

    print(f"[image] 下载 {image_url[:80]}…", flush=True)
    try:
        resp = requests.get(image_url, timeout=timeout, stream=True)
        resp.raise_for_status()
        content_type = resp.headers.get("content-type", "")
        if "image" not in content_type:
            preview = resp.text[:200] if resp.text else "(empty)"
            print(f"[image] ⚠️ 响应不是图片 ({content_type}): {preview}")
            return ""
        total = 0
        with open(local_path, "wb") as f:
            for chunk in resp.iter_content(chunk_size=8192):
                f.write(chunk)
                total += len(chunk)
        if total == 0:
            local_path.unlink(missing_ok=True)
            return ""
        print(f"[image] ✅ 下载完成 ({total} bytes) → {local_path}", flush=True)
        return str(local_path.absolute())
    except Exception as e:
        print(f"[image] 下载失败: {e}")
        try:
            local_path.unlink(missing_ok=True)
        except Exception:
            pass
        return ""


def generate_product_image(
    config: Config,
    user: UserRecord,
    product_image_path: str | None = None,
    skip: bool = False,
    max_retries: int = 3,
    retry_delay: int = 12,
    output_dir: str = "output/images",
) -> str:
    """生成营销图片主入口

    Args:
        product_image_path: 如果有现成的本地产品图，直接跳过生成走叠加层。
        skip: 为 True 时完全不生成图片（仅测文案/HTML 用）。
    返回：本地图片路径（绝对）或空串
    """
    if skip:
        return ""

    # 已有本地产品图 → 直接叠加文字（极快）
    if product_image_path and os.path.exists(product_image_path):
        if getattr(config.marketing, "overlay_text", True):
            return overlay_marketing_text(
                image_path=product_image_path,
                discount=user.discount,
                brand_name=user.brand,
                cta_text=config.marketing.cta_button,
            )
        return str(Path(product_image_path).absolute())

    prompt = generate_image_prompt(user, config)
    wanx_ready = bool(config.qianwen_vision.api_key and config.qianwen_vision.base_url)

    for attempt in range(1, max_retries + 1):
        try:
            if wanx_ready:
                print(f"[image] 第 {attempt}/{max_retries} 次，调用万相生成…", flush=True)
                image_url = _request_wanx(prompt, config)
            else:
                # 万相未配置：Pollinations 兜底 + 必做文字叠加（因为通用图里没有折扣/CTA）
                print(f"[image] 第 {attempt}/{max_retries} 次，Pollinations 兜底（overlay=True）…", flush=True)
                image_url = _request_pollinations(user.discount, user.brand, seed_salt=str(attempt))

            if image_url.startswith("/") or image_url.startswith("\\") or "://" not in image_url:
                local = image_url
            else:
                local = download_image(image_url, user.user_id, output_dir=output_dir)

            if not local:
                raise RuntimeError("图片下载为空")

            # 万相的 overlay_text 默认 False（文字已由模型画进图）；Pollinations 强制 True
            need_overlay = (
                True
                if not wanx_ready
                else bool(getattr(config.marketing, "overlay_text", False))
            )
            if need_overlay:
                final = overlay_marketing_text(
                    image_path=local,
                    discount=user.discount,
                    brand_name=user.brand,
                    cta_text=config.marketing.cta_button,
                )
                return final
            return str(Path(local).absolute())

        except Exception as e:
            print(f"[image] 第 {attempt}/{max_retries} 次失败: {e}", flush=True)
            if attempt < max_retries:
                time.sleep(retry_delay)

    print("[image] 达到最大重试次数，返回空", flush=True)
    return ""

File: backend/ts/image-generator.ts
/**
 * 图片生成模块 — 双轨策略
 *  1. 万相（wanx）：config.qianwen_vision 配 api_key+base_url 时调阿里万相文生图
 *  2. Pollinations 兜底：免 Key，通用图 + Pillow（→ @napi-rs/canvas）叠加折扣/CTA
 * 端口自 Python emailgen/image_generator.py。
 */
import * as fs from 'fs';
import * as path from 'path';

import type { Config } from './config';
import type { UserRecord } from './data-loader';
import { generateImagePrompt } from './copy-generator';
import { overlayMarketingText } from './image-overlay';

function extractImageUrl(data: Record<string, unknown>): string {
  const out = (data.output || {}) as Record<string, unknown>;
  const choices = (out.choices || data.choices || []) as unknown[];
  if (!choices.length) return '';
  const choice = choices[0] as Record<string, unknown>;
  const msg = (choice.message || {}) as Record<string, unknown>;
  const content = msg.content;

  if (Array.isArray(content)) {
    for (const part of content) {
      if (!part || typeof part !== 'object') continue;
      const p = part as Record<string, unknown>;
      if (p.image) return String(p.image);
      if (p.type === 'image_url' && p.image_url && typeof p.image_url === 'object') {
        return String((p.image_url as Record<string, unknown>).url ?? '');
      }
      if (p.url) return String(p.url);
    }
    return '';
  }

  if (typeof content === 'string') {
    const m = content.match(/\((https?:\/\/[^)]+)\)/) || content.match(/(https?:\/\/\S+)/);
    return m ? m[1] : '';
  }
  return '';
}

async function fetchWithTimeout(url: string, init: RequestInit, timeoutMs: number): Promise<Response> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    return await fetch(url, { ...init, signal: controller.signal });
  } finally {
    clearTimeout(timer);
  }
}

async function requestWanx(prompt: string, config: Config): Promise<string> {
  const q = config.qianwen_vision;
  const base = (q.base_url || '').replace(/\/$/, '');
  if (!base || !q.api_key) throw new Error('万相未配置（缺少 base_url 或 api_key）');
  const url = `${base}/chat/completions`;
  const headers = { Authorization: `Bearer ${q.api_key}`, 'Content-Type': 'application/json' };
  const payload = {
    model: q.model,
    messages: [{ role: 'user', content: [{ type: 'text', text: prompt }] }],
  };
  const resp = await fetchWithTimeout(
    url,
    { method: 'POST', headers, body: JSON.stringify(payload) },
    180000,
  );
  if (resp.status !== 200) throw new Error(`万相 HTTP ${resp.status}: ${(await resp.text()).slice(0, 400)}`);
  const data = (await resp.json()) as Record<string, unknown>;
  const imageUrl = extractImageUrl(data);
  if (!imageUrl) throw new Error(`万相响应无图片 URL: ${JSON.stringify(data).slice(0, 400)}`);
  return imageUrl;
}

function stableHash(s: string): number {
  let h = 0;
  for (let i = 0; i < s.length; i++) h = (Math.imul(31, h) + s.charCodeAt(i)) | 0;
  return Math.abs(h);
}

function requestPollinations(discount: number, brand: string, seedSalt = ''): string {
  const prompt =
    'High-quality e-commerce marketing email hero image of a premium phone case, ' +
    'sleek modern smartphone case product photography, ' +
    'bright clean studio lighting, soft gradient background, ' +
    'professional brand aesthetic, newsletter banner style, ' +
    '8K detailed, no text overlay in the image, best quality';
  const encoded = encodeURIComponent(prompt);
  const seed =
    (Math.floor(Date.now() / 1000) + (stableHash(`${brand}_${discount}_${seedSalt}`) % 100000)) % 1000000;
  return `https://image.pollinations.ai/prompt/${encoded}?width=1200&height=1500&seed=${seed}&nologo=true`;
}

export async function downloadImage(
  imageUrl: string,
  tag: string,
  outputDir = 'output/images',
  timeout = 90,
): Promise<string> {
  const outDir = path.isAbsolute(outputDir) ? outputDir : path.resolve(process.cwd(), outputDir);
  fs.mkdirSync(outDir, { recursive: true });
  const safeTag = (tag.replace(/[^\w.-]+/g, '_').slice(0, 60) || 'img');
  const filename = `${safeTag}_${Math.floor(Date.now() / 1000)}.png`;
  const localPath = path.join(outDir, filename);

  console.error(`[image] 下载 ${imageUrl.slice(0, 80)}…`);
  try {
    const resp = await fetchWithTimeout(imageUrl, {}, timeout * 1000);
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    const contentType = resp.headers.get('content-type') || '';
    if (!contentType.includes('image')) {
      const text = await resp.text();
      console.error(`[image] ⚠️ 响应不是图片 (${contentType}): ${text.slice(0, 200)}`);
      return '';
    }
    const buf = Buffer.from(await resp.arrayBuffer());
    if (buf.length === 0) {
      try { fs.unlinkSync(localPath); } catch { /* ignore */ }
      return '';
    }
    fs.writeFileSync(localPath, buf);
    console.error(`[image] ✅ 下载完成 (${buf.length} bytes) → ${localPath}`);
    return path.resolve(localPath);
  } catch (e) {
    console.error(`[image] 下载失败: ${(e as Error).message}`);
    try { fs.unlinkSync(localPath); } catch { /* ignore */ }
    return '';
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

export interface GenerateImageOpts {
  config: Config;
  user: UserRecord;
  productImagePath?: string | null;
  skip?: boolean;
  maxRetries?: number;
  retryDelay?: number;
  outputDir?: string;
}

export async function generateProductImage(opts: GenerateImageOpts): Promise<string> {
  const { config, user } = opts;
  const skip = opts.skip ?? false;
  const productImagePath = opts.productImagePath ?? null;
  const maxRetries = opts.maxRetries ?? 3;
  const retryDelay = opts.retryDelay ?? 12;
  const outputDir = opts.outputDir ?? 'output/images';

  if (skip) return '';

  // 已有本地产品图 → 直接叠加（或原样返回）
  if (productImagePath && fs.existsSync(productImagePath)) {
    if (config.marketing.overlay_text) {
      return overlayMarketingText({
        imagePath: productImagePath,
        discount: user.discount,
        brandName: user.brand,
        ctaText: config.marketing.cta_button,
        ageRange: user.age_range,
      });
    }
    return path.resolve(productImagePath);
  }

  const prompt = generateImagePrompt(user, config);
  const wanxReady = Boolean(config.qianwen_vision.api_key && config.qianwen_vision.base_url);

  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      let imageUrl: string;
      if (wanxReady) {
        console.error(`[image] 第 ${attempt}/${maxRetries} 次，调用万相生成…`);
        imageUrl = await requestWanx(prompt, config);
      } else {
        console.error(`[image] 第 ${attempt}/${maxRetries} 次，Pollinations 兜底（overlay=true）…`);
        imageUrl = requestPollinations(user.discount, user.brand, String(attempt));
      }

      let local: string;
      if (imageUrl.startsWith('/') || imageUrl.startsWith('\\') || !imageUrl.includes('://')) {
        local = imageUrl;
      } else {
        local = await downloadImage(imageUrl, user.user_id, outputDir);
      }
      if (!local) throw new Error('图片下载为空');

      const needOverlay = !wanxReady ? true : Boolean(config.marketing.overlay_text);
      if (needOverlay) {
        const final = await overlayMarketingText({
          imagePath: local,
          discount: user.discount,
          brandName: user.brand,
          ctaText: config.marketing.cta_button,
          ageRange: user.age_range,
        });
        return final;
      }
      return path.resolve(local);
    } catch (e) {
      console.error(`[image] 第 ${attempt}/${maxRetries} 次失败: ${(e as Error).message}`);
      if (attempt < maxRetries) await sleep(retryDelay * 1000);
    }
  }
  console.error('[image] 达到最大重试次数，返回空');
  return '';
}
